test/infrastructure/inmemory/internal/server/mux.go

```diff
@@ -316,81 +318,104 @@ func (m *WorkloadClustersMux) initWorkloadClusterListenerWithPortLocked(wclName
-		if err != nil {
-			return errors.Wrapf(err, "failed to create serving certificate for API server %s", podName)
+		// Generate admin certificates to be used for accessing the API server.
+		// NOTE: this is used for tests because CAPI creates its own.
+		if wcl.adminCertificate == nil {
+			config := adminClientCertificateConfig()
+			cert, key, err := newCertAndKey(caCert, caKey, config)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create admin certificate for API server %s", podName)
+			}
+
+			wcl.adminCertificate = cert
+			wcl.adminKey = key
 		}
 
-		certificate, err := tls.X509KeyPair(certs.EncodeCertPEM(cert), certs.EncodePrivateKeyPEM(key))
-		if err != nil {
-			return errors.Wrapf(err, "failed to create X509KeyPair for API server %s", podName)
+		// Start the listener for the API server.
+		// NOTE: There is only one listener for all API server instances; the same listener will act
+		// as a port forward target too.
+		if wcl.listener != nil {
+			return nil
 		}
-		wcl.apiServerServingCertificate = &certificate
-	}
 
-	// Generate admin certificates to be used for accessing the API server.
-	// NOTE: this is used for tests because CAPI creates its own.
-	if wcl.adminCertificate == nil {
-		config := adminClientCertificateConfig()
-		cert, key, err := newCertAndKey(caCert, caKey, config)
+		l, err := net.Listen("tcp", wcl.HostPort())
 		if err != nil {
-			return errors.Wrapf(err, "failed to create admin certificate for API server %s", podName)
+			return errors.Wrapf(err, "failed to start WorkloadClusterListener %s, %s", wclName, wcl.HostPort())
 		}
+		wcl.listener = l
 
-		wcl.adminCertificate = cert
-		wcl.adminKey = key
-	}
-
-	// Start the listener for the API server.
-	// NOTE: There is only one listener for all API server instances; the same listener will act
-	// as a port forward target too.
-	if wcl.listener != nil {
+		go func() {
+			if startServerErr = m.muxServer.ServeTLS(wcl.listener, "", ""); startServerErr != nil && !errors.Is(startServerErr, http.ErrServerClosed) {
+				m.log.Error(startServerErr, "Failed to start WorkloadClusterListener", "listenerName", wclName, "address", wcl.Address())
+			}
+		}()
 		return nil
-	}
+	}()
 
-	l, err := net.Listen("tcp", wcl.HostPort())
-	if err != nil {
-		return errors.Wrapf(err, "failed to start WorkloadClusterListener %s, %s", wclName, wcl.HostPort())
-	}
-	wcl.listener = l
-
-	var startErr error
-	startCh := make(chan struct{})
-	go func() {
-		startCh <- struct{}{}
-		if err := m.muxServer.ServeTLS(wcl.listener, "", ""); err != nil && !errors.Is(err, http.ErrServerClosed) {
-			startErr = err
-			m.log.Error(startErr, "Failed to start WorkloadClusterListener", "listenerName", wclName, "address", wcl.Address())
+	// Wait until the sever is working.
+	var pollErr error
+	err = wait.PollUntilContextTimeout(context.TODO(), 10*time.Millisecond, 1*time.Second, true, func(ctx context.Context) (done bool, err error) {
```

Comment:
With this change we wait until the listener serves (at least a tls handshake). This is quicker than always waiting 100 ms